### PR TITLE
[CodeGen] Add a pattern to canonicalize HAL interface load/store

### PR DESCRIPTION
**Pull request by antiagainst**
This pass adds one pattern that folds tensor reshapes into the
loaded subspan. Along the way, renamed RemoveDeadMemAllocsPass
into BufferAllocViewCleanUpPass.

After HAL interface materialization, outlined dispatch regions
become modules to be compiled into executables run on a specific
target. Those modules uses HAL interface binding ops to access
buffers holding input/output data. From IREE's perspective,
buffers just contain a bag of bytes. So the executable translation
is free to reinterpret the buffer in whatever shape that fits.
Therefore we can fold reshapes.

This enables turning on depthwise convolution vectorization patterns
in SPIR-V CodeGen path.

### CODE REVIEW

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Looks good overall, but why separate pass? I am weary having passes for specific things. This seems like it could just be a canonicalization. We are planning to use linalg on tensors as the front end for IREE anyway....

**Comment by antiagainst** [comment]
> Looks good overall, but why separate pass? I am weary having passes for specific things. This seems like it could just be a canonicalization. We are planning to use linalg on tensors as the front end for IREE anyway....

Yeah. I was debating for a while where to put this actually. It's a pattern that involves 3 dialects: Flow/HAL/Linalg. It's matching against a `linalg.tensor_reshape` op. If we want to set this as a canonicalization pattern, it should be on `linalg.tensor_reshape` ops. That clearly cannot happen. Making it a canonoicalization pattern for `flow.dispatch.tensor.load` or `hal.interface.binding.subspan` ops means we need to check backwards on their uses, which is contradictory to what MLIR patterns generally do too. And it means pulling in 2 dialects as dependencies. All these sounds not proper to me. So a separate pass.

Taking another look of existing passes, we actually have another pass, `RemoveDeadMemAllocsPass`, that also just contain on simple pattern. I think one way to save yet another pass might be putting the pattern here into it and rename that pass as `BufferAllocViewCleanupPass`, given they are all performing clean ups on buffer allocations/views. We can collect more patterns in that pass later if more pop up. WDTY?

**Comment by MaheshRavishankar** [comment]
> Taking another look of existing passes, we actually have another pass, `RemoveDeadMemAllocsPass`, that also just contain on simple pattern. I think one way to save yet another pass might be putting the pattern here into it and rename that pass as `BufferAllocViewCleanupPass`, given they are all performing clean ups on buffer allocations/views. We can collect more patterns in that pass later if more pop up. WDTY?

Sounds good to me.

File: iree/compiler/Conversion/Common/BufferAllocViewCleanUpPass.cpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    : OpRewritePattern<linalg::TensorReshapeOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::TensorReshapeOp reshapeOp,
+                                PatternRewriter &rewriter) const override {
+    auto loadOp =
+        reshapeOp.src().getDefiningOp<IREE::Flow::DispatchTensorLoadOp>();
+    if (!loadOp) return failure();
+
+    if (!loadOp.offsets().empty() || !loadOp.sizes().empty() ||
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/Common/BufferAllocViewCleanUpPass.cpp:65]: Nit: Merge this if with the others.

**Review by MaheshRavishankar — APPROVED** [review]

**Pull request merged**